amr-wind/boundary_conditions/MassInflowOutflowBC.cpp

```diff
@@ -0,0 +1,75 @@
+
+#include "amr-wind/boundary_conditions/MassInflowOutflowBC.H"
+
+namespace amr_wind {
+namespace {
+
+} // namespace
+
+MassInflowOutflowBC::MassInflowOutflowBC(Field& field, amrex::Orientation ori)
+    : m_field(field), m_ori(ori)
+{}
+
+void MassInflowOutflowBC::operator()(Field& /*field*/, const FieldState /*rho_state*/)
+{
+    const auto& repo = m_field.repo();
+    const auto& velocity = repo.get_field("velocity");
+    //const auto bcvals = m_field.bc_values_device();
+    const int ncomp = m_field.num_comp();
+    const int idx = static_cast<int>(m_ori);
+    const int idim = m_ori.coordDir();
+    const auto islow = m_ori.isLow();
+    const auto ishigh = m_ori.isHigh();
+    const int nlevels = m_field.repo().num_active_levels();
+    const bool ib = (idim == 0), jb = (idim == 1), kb = (idim == 2);
+
+    amrex::Print() << "***** Applying MIO custom Neumann fills at orientation: " << idx << std::endl;
+    for (int lev = 0; lev < nlevels; ++lev) {
+        const auto& domain = repo.mesh().Geom(lev).Domain();
+
+        amrex::MFItInfo mfi_info{};
+        if (amrex::Gpu::notInLaunchRegion()) {
+            mfi_info.SetDynamic(true);
+        }
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi(m_field(lev), mfi_info); mfi.isValid(); ++mfi) {
+            auto bx = mfi.validbox();
+            //bx.grow({!ib, !jb, !kb});   // how to manage corner cells??
+            const auto& bc_a = m_field(lev).array(mfi);
+            const auto& vel = velocity(lev).array(mfi);
+
+            if (islow && (bx.smallEnd(idim) == domain.smallEnd(idim))) {
+                amrex::ParallelFor(
+                    amrex::bdryLo(bx, idim),
+                    [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        if (vel(i-ib, j-jb, k-kb, idim) < 0) {
+                            for (int n = 0; n < ncomp; n++) {
```

Comment:
I believe that wouldn't work because we want to check the if condition (for outflow) on the original boundary value of velocity. For instance, we would want to check the the x-velocity at the x-boundary for the outflow, and if it is outflow, it is overwritten by the first internal cell. But then in the next loop iteration which operates on the y-velocity, we longer have the original x-velocity to check for outflow. I guess that's why the FixedGradientBC also has a separate component loop. I naively merged it with the ParallelFor when I started but figured out the above issue.